File: ulyp-ui/src/main/java/com/ulyp/ui/code/find/JarFile.java
package com.ulyp.ui.code.find;

import com.ulyp.ui.code.SourceCode;
import com.ulyp.ui.util.StreamDrainer;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.util.zip.ZipEntry;

public class JarFile {

    private static final String CLASS_FILE_EXTENSION = ".class";
    private static final String CLASS_SOURCE_FILE_EXTENSION = ".java";

    private final File file;
    private final java.util.jar.JarFile jarFile;

    protected JarFile(File file) {
        try {
            this.file = file;
            this.jarFile = new java.util.jar.JarFile(file.getAbsolutePath());
        } catch (IOException e) {
            throw new RuntimeException("Could not find jar file: " + file, e);
        }
    }

    /**
     * For maven/gradle jar files it's possible to locate jar file with source code
     * @return jar file with source if possible to locate or null otherwise
     */
    @Nullable
    public JarFile deriveSourcesJar() {
        Path libFolder = file.toPath().getParent().getParent();

        for (File p : libFolder.toFile().listFiles()) {
            if (p.isDirectory()) {
                for (File jarFile : p.listFiles()) {
                    if (jarFile.getAbsolutePath().contains("-sources.jar")) {
                        return new JarFile(jarFile);
                    }
                }
            }
        }

        return null;
    }

    @Nullable
    public SourceCode findSourceByClassName(String className) {
        ZipEntry zipEntry = jarFile.getEntry(className.replace('.', '/') + CLASS_SOURCE_FILE_EXTENSION);
        if (zipEntry == null) {
            return null;
        } else {
            try (InputStream inputStream = jarFile.getInputStream(zipEntry)) {
                return new SourceCode(className, new String(StreamDrainer.DEFAULT.drain(inputStream)));
            } catch (IOException e) {
                throw new RuntimeException("Could not read jar file: " + e.getMessage(), e);
            }
        }
    }

    @Nullable
    public ByteCode findByteCodeByClassName(String className) {
        ZipEntry zipEntry = jarFile.getEntry(className.replace('.', '/') + CLASS_FILE_EXTENSION);
        if (zipEntry == null) {
            return null;
        } else {
            try (InputStream inputStream = jarFile.getInputStream(zipEntry)) {
                return new ByteCode(className, StreamDrainer.DEFAULT.drain(inputStream));
            } catch (IOException e) {
                throw new RuntimeException("Could not read jar file: " + e.getMessage(), e);
            }
        }
    }

    public String getAbsolutePath() {
        return this.file.getAbsolutePath();
    }
}

File: ulyp-ui/src/main/kotlin/com/ulyp/ui/code/find/JarFile.kt
package com.ulyp.ui.code.find

import com.ulyp.ui.code.SourceCode
import com.ulyp.ui.util.StreamDrainer
import java.io.File
import java.io.IOException

open class JarFile(private val file: File) {
    private var jarFile: java.util.jar.JarFile

    companion object {
        private const val CLASS_FILE_EXTENSION = ".class"
        private const val CLASS_SOURCE_FILE_EXTENSION = ".java"
    }

    init {
        try {
            jarFile = java.util.jar.JarFile(file.absolutePath)
        } catch (e: IOException) {
            throw RuntimeException("Could not find jar file: $file", e)
        }
    }

    /**
     * For maven/gradle jar files it's possible to locate jar file with source code
     * @return jar file with source if possible to locate or null otherwise
     */
    fun deriveSourcesJar(): JarFile? {
        val libFolder = file.toPath().parent.parent
        for (p in libFolder.toFile().listFiles()) {
            if (p.isDirectory) {
                for (jarFile in p.listFiles()) {
                    if (jarFile.absolutePath.contains("-sources.jar")) {
                        return JarFile(jarFile)
                    }
                }
            }
        }
        return null
    }

    fun findSourceByClassName(className: String): SourceCode? {
        val zipEntry = jarFile.getEntry(className.replace('.', '/') + CLASS_SOURCE_FILE_EXTENSION)
        if (zipEntry == null) {
            return null
        } else {
            try {
                jarFile.getInputStream(zipEntry).use { inputStream ->
                    return SourceCode(
                        className,
                        String(StreamDrainer.DEFAULT.drain(inputStream))
                    )
                }
            } catch (e: IOException) {
                throw RuntimeException("Could not read jar file: " + e.message, e)
            }
        }
    }

    fun findByteCodeByClassName(className: String): ByteCode? {
        val zipEntry = jarFile.getEntry(className.replace('.', '/') + CLASS_FILE_EXTENSION)
        if (zipEntry == null) {
            return null
        } else {
            try {
                jarFile.getInputStream(zipEntry)
                    .use { inputStream -> return ByteCode(className, StreamDrainer.DEFAULT.drain(inputStream)) }
            } catch (e: IOException) {
                throw RuntimeException("Could not read jar file: " + e.message, e)
            }
        }
    }

    val absolutePath: String
        get() = file.absolutePath
}